velero-plugins/imagestreamtag/backup.go

```diff
@@ -0,0 +1,92 @@
+package imagestreamtag
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/konveyor/openshift-velero-plugin/velero-plugins/clients"
+	"github.com/konveyor/openshift-velero-plugin/velero-plugins/common"
+	imagev1API "github.com/openshift/api/image/v1"
+	"github.com/sirupsen/logrus"
+	v1 "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	"github.com/vmware-tanzu/velero/pkg/plugin/velero"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+)
+
+// BackupPlugin is a backup item action plugin for Velero
+type BackupPlugin struct {
+	Log logrus.FieldLogger
+}
+
+// AppliesTo returns a velero.ResourceSelector that applies to imagestreamtags
+func (p *BackupPlugin) AppliesTo() (velero.ResourceSelector, error) {
+	return velero.ResourceSelector{
+		IncludedResources: []string{"imagestreamtags"},
+	}, nil
+}
+
+// Execute sets annotations on imagestreamtags that depend on others to be restored first
+func (p *BackupPlugin) Execute(item runtime.Unstructured, backup *v1.Backup) (runtime.Unstructured, []velero.ResourceIdentifier, error) {
+
+	p.Log.Info("[istag-backup] Entering ImageStreamTag backup plugin")
+	imageStreamTag := imagev1API.ImageStreamTag{}
+	itemMarshal, _ := json.Marshal(item)
+	json.Unmarshal(itemMarshal, &imageStreamTag)
+	annotations := imageStreamTag.Annotations
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+	// clear out any previous istag annotations from old migrations
+	if _, found := annotations[common.RelatedIsTagAnnotation]; found {
+		delete(annotations, common.RelatedIsTagAnnotation)
+	}
+	if _, found := annotations[common.RelatedIsTagNsAnnotation]; found {
+		delete(annotations, common.RelatedIsTagNsAnnotation)
+	}
+	p.Log.Info(fmt.Sprintf("[istag-backup] Backing up imagestreamtag %s", imageStreamTag.Name))
+
+	referenceTag := imageStreamTag.Tag != nil && imageStreamTag.Tag.From != nil
+	if referenceTag {
+		p.Log.Info(fmt.Sprintf("[istag-backup] Reference tag: %v, tag: %v", imageStreamTag.Tag.From.Kind, imageStreamTag.Tag.From.Name))
+
+		tagNamespace := imageStreamTag.Tag.From.Namespace
+		if tagNamespace == "" {
+			tagNamespace = imageStreamTag.Namespace
+		}
+		client, err := clients.ImageClient()
+		if imageStreamTag.Tag.From.Kind == "ImageStreamTag" {
+			p.Log.Info(fmt.Sprintf("[istag-backup] Looking up reference tag: %s/%s", tagNamespace, imageStreamTag.Tag.From.Name))
+			_, err = client.ImageStreamTags(tagNamespace).Get(imageStreamTag.Tag.From.Name, metav1.GetOptions{})
+			if err == nil {
```

Comment:
What happens when err isn't nil here?

Reply:
It means we can't find the associated istag in the src cluster, which means there's no way to set the annotation to tell the restore plugin to restore the associated istag before this one.
Since the only purpose of this annotation is to help to prevent dest/restore-cluster race conditions surrounding istag restore order, there's nothing for us to do if the associated istag doesn't exist.